File: ts/scopes/parser.ts
export interface Scope {
  resource: string;
  action: string;
  qualifier?: string;
  modifiers: Record<string, string | boolean>;
}

const SCOPE_REGEX =
  /^(?<res>[A-Z_]+)\.(?<act>[A-Z_]+)(:(?<qual>[^()]+))?(\((?<mod>.+)\))?$/;

export function parseScope(input: string): Scope {
  const match = input.match(SCOPE_REGEX);
  if (!match?.groups) throw new Error(`Invalid scope: ${input}`);

  const mods: Record<string, string | boolean> = {};
  if (match.groups.mod) {
    match.groups.mod.split(",").forEach(pair => {
      const [k, v] = pair.split(":");
      mods[k.trim()] = v ? v.trim() : true;
    });
  }

  return {
    resource: match.groups.res,
    action: match.groups.act,
    qualifier: match.groups.qual,
    modifiers: mods
  };
}

File: py/scopes/parser.py
import re

SCOPE_RE = re.compile(
    r'^(?P<res>[A-Z_]+)\.(?P<act>[A-Z_]+)'
    r'(:(?P<qual>[^()]+))?'
    r'(\((?P<mod>.+)\))?$'
)

def parse_scope(scope: str):
    m = SCOPE_RE.match(scope)
    if not m:
        raise ValueError(f"Invalid scope: {scope}")

    mods = {}
    if m.group("mod"):
        for part in m.group("mod").split(","):
            if ":" in part:
                k, v = part.split(":")
                mods[k.strip()] = v.strip()
            else:
                mods[part.strip()] = True

    return {
        "resource": m.group("res"),
        "action": m.group("act"),
        "qualifier": m.group("qual"),
        "modifiers": mods
    }